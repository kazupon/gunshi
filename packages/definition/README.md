# `@gunshi/definition`

[![Version][npm-version-src]][npm-version-href]
[![InstallSize][install-size-src]][install-size-src]
[![JSR][jsr-src]][jsr-href]

> utilities for gunshi command definition

This package exports the bellow APIs and types.

- `define`: A function to define a command.
- `lazy`: A function to lazily load a command.
- Some basic type definitions, such as `Command`, `LazyCommand`, etc.

<!-- eslint-disable markdown/no-missing-label-refs -->

> [!TIP]
> **The APIs and type definitions available in this package are the same as those in the `gunshi/definition` entry in the `gunshi` package.**
> This package is smaller in file size than the `gunshi` package, making it suitable for use when you want to reduce the size of the `node_modules` in your command you are creating.

<!-- eslint-enable markdown/no-missing-label-refs -->

## üíø Installation

```sh
# npm
npm install --save @gunshi/definition

## pnpm
pnpm add @gunshi/definition

## yarn
yarn add @gunshi/definition

## deno
deno add jsr:@gunshi/definition

## bun
bun add @gunshi/definition
```

## üöÄ Usage

You can define the gunshi command as JavaScript module with using `define`, `defineWithTypes`, `lazy` and `lazyWithTypes`

The bellow example case which is using `define`:

```js
import { define } from '@gunshi/definition'

/**
 * define `say` command as javascript module
 */
export default define({
  name: 'say',
  args: {
    say: {
      type: 'string',
      description: 'say something',
      default: 'hello!'
    }
  },
  run: ctx => {
    return `You said: ${ctx.values.say}`
  }
})
```

## ¬©Ô∏è License

[MIT](http://opensource.org/licenses/MIT)

<!-- Badges -->

[npm-version-src]: https://img.shields.io/npm/v/@gunshi/definition?style=flat
[npm-version-href]: https://npmjs.com/package/@gunshi/definition@alpha
[jsr-src]: https://jsr.io/badges/@gunshi/definition
[jsr-href]: https://jsr.io/@gunshi/definition
[install-size-src]: https://pkg-size.dev/badge/install/23122
