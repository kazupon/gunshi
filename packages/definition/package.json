{
  "name": "@gunshi/definition",
  "description": "utilities for gunshi command definition",
  "version": "0.27.0-alpha.8",
  "author": {
    "name": "kazuya kawaguchi",
    "email": "kawakazu80@gmail.com"
  },
  "license": "MIT",
  "funding": "https://github.com/sponsors/kazupon",
  "bugs": {
    "url": "https://github.com/kazupon/gunshi/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/kazupon/gunshi.git",
    "directory": "packages/definition"
  },
  "keywords": [
    "gunshi",
    "command",
    "definition",
    "cli"
  ],
  "publishConfig": {
    "access": "public"
  },
  "engines": {
    "node": ">= 20"
  },
  "type": "module",
  "files": [
    "lib"
  ],
  "module": "lib/index.js",
  "exports": {
    ".": {
      "types": "./lib/index.d.ts",
      "import": "./lib/index.js",
      "require": "./lib/index.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "types": "lib/index.d.ts",
  "typesVersions": {
    "*": {
      "*": [
        "./lib/*",
        "./*"
      ]
    }
  },
  "scripts": {
    "build": "tsdown",
    "lint:jsr": "jsr publish --dry-run --allow-dirty",
    "prepack": "pnpm build",
    "typecheck:deno": "deno check --import-map=../../importmap.json ./src"
  },
  "devDependencies": {
    "deno": "catalog:",
    "gunshi": "workspace:*",
    "jsr": "catalog:",
    "jsr-exports-lint": "catalog:",
    "publint": "catalog:",
    "tsdown": "catalog:"
  }
}
