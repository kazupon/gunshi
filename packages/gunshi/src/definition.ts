/**
 * The entry for command definition.
 *
 * @example
 * ```js
 * import { define } from 'gunshi/definition'
 * ```
 *
 * @module
 */

/**
 * @author kazuya kawaguchi (a.k.a. kazupon)
 * @license MIT
 */

import type { Args } from 'args-tokens'
import type {
  Awaitable,
  Command,
  CommandContext,
  CommandContextExt,
  CommandContextExtension,
  CommandExamplesFetcher,
  CommandLoader,
  CommandResourceFetcher,
  CommandRunner,
  ExtendedCommand,
  LazyCommand
} from './types.ts'

export type { Args, ArgSchema, ArgValues } from 'args-tokens'

// Overload for command without extensions
export function define<A extends Args = Args>(
  definition: {
    name?: string
    description?: string
    args?: A
    examples?: string | CommandExamplesFetcher<A>
    resource?: CommandResourceFetcher<A>
    toKebab?: boolean
    run?: CommandRunner<A>
  } & { extensions?: never }
): Command<A>

// Overload for command with extensions
export function define<
  A extends Args = Args,
  E extends Record<string, CommandContextExtension> = Record<string, CommandContextExtension>
>(definition: {
  name?: string
  description?: string
  args?: A
  examples?: string | CommandExamplesFetcher<A>
  resource?: CommandResourceFetcher<A>
  toKebab?: boolean
  extensions: E
  run?: (ctx: Readonly<CommandContext<A> & CommandContextExt<E>>) => Awaitable<void | string>
}): ExtendedCommand<A, E>

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function define(definition: any): any {
  if ('extensions' in definition && definition.extensions) {
    const { extensions, ...rest } = definition
    return {
      ...rest,
      _extensions: extensions
    }
  }
  return definition
}

/**
 * Define a {@link LazyCommand | lazy command} without definition
 * @param loader A {@link CommandLoader | command loader}
 * @returns A {@link LazyCommand | lazy command} loader
 */
export function lazy<A extends Args = Args>(
  loader: CommandLoader<A>
): LazyCommand<A> & { _extensions?: never }

/**
 * Define a {@link LazyCommand | lazy command} with command loader, which is attached with command definition as usage metadata.
 * @param loader A {@link CommandLoader | command loader}
 * @param definition A {@link Command | command} definition
 * @returns A {@link LazyCommand | lazy command} loader
 */
export function lazy<A extends Args = Args>(
  loader: CommandLoader<A>,
  definition: Omit<Command<A>, 'run'>
): LazyCommand<A> & { _extensions?: never }

/**
 * Define a {@link LazyCommand | lazy command} with extension support (extensions property)
 * @param loader A {@link CommandLoader | command loader}
 * @param definition An {@link ExtendedCommand | extended command} definition with extensions
 * @returns A {@link LazyCommand | lazy command} loader with extension support
 */
export function lazy<
  A extends Args = Args,
  E extends Record<string, CommandContextExtension> = Record<string, CommandContextExtension>
>(
  loader: CommandLoader<A>,
  definition: Omit<Command<A>, 'run'> & { extensions: E }
): LazyCommand<A> & { _extensions: E }

/**
 * Define a {@link LazyCommand | lazy command} with extension support (_extensions property)
 * @param loader A {@link CommandLoader | command loader}
 * @param definition An {@link ExtendedCommand | extended command} definition with _extensions
 * @returns A {@link LazyCommand | lazy command} loader with extension support
 */
export function lazy<
  A extends Args = Args,
  E extends Record<string, CommandContextExtension> = Record<string, CommandContextExtension>
>(
  loader: CommandLoader<A>,
  definition: Omit<Command<A>, 'run'> & { _extensions: E }
): LazyCommand<A> & { _extensions: E }

export function lazy<
  A extends Args = Args,
  E extends Record<string, CommandContextExtension> = Record<string, CommandContextExtension>
>(
  loader: CommandLoader<A>,
  definition?: Omit<Command<A>, 'run'> & ({ extensions?: E } | { _extensions?: E })
): LazyCommand<A> & { _extensions?: E | never } {
  const lazyCommand = loader as LazyCommand<A> & { _extensions?: E | never }

  if (definition != null) {
    // copy existing properties
    lazyCommand.commandName = definition.name
    lazyCommand.description = definition.description
    lazyCommand.args = definition.args
    lazyCommand.examples = definition.examples
    lazyCommand.resource = definition.resource
    lazyCommand.toKebab = definition.toKebab

    // handle extensions
    if ('extensions' in definition && definition.extensions) {
      lazyCommand._extensions = definition.extensions
    } else if ('_extensions' in definition && definition._extensions) {
      lazyCommand._extensions = definition._extensions
    }
  }

  return lazyCommand
}
