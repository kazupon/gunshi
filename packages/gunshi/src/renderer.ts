/**
 * The entry point for Gunshi renderer.
 *
 * @example
 * ```js
 * import { renderHeader, renderUsage, renderValidationErrors } from 'gunshi/renderer'
 * ```
 *
 * @module
 */

/**
 * @author kazuya kawaguchi (a.k.a. kazupon)
 * @license MIT
 */

export { renderHeader } from './plugins/renderer/header.ts'
export { renderUsage } from './plugins/renderer/usage.ts'
export { renderValidationErrors } from './plugins/renderer/validation.ts'
