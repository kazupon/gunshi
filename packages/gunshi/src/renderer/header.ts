/**
 * @author kazuya kawaguchi (a.k.a. kazupon)
 * @license MIT
 */

import type { Args } from 'args-tokens'
import type { CommandContext } from '../types.ts'

/**
 * Render the header.
 * @param ctx A {@link CommandContext | command context}
 * @returns A rendered header.
 */
export function renderHeader<A extends Args = Args>(
  ctx: Readonly<CommandContext<A>>
): Promise<string> {
  const title = ctx.env.description || ctx.env.name || ''
  return Promise.resolve(
    title
      ? `${title} (${ctx.env.name || ''}${ctx.env.version ? ` v${ctx.env.version}` : ''})`
      : title
  )
}
