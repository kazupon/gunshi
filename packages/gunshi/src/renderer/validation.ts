/**
 * @author kazuya kawaguchi (a.k.a. kazupon)
 * @license MIT
 */

import type { CommandContext, DefaultGunshiParams, GunshiParams } from '../types.ts'

/**
 * Render the validation errors.
 * @param ctx A {@link CommandContext | command context}
 * @param error An {@link AggregateError} of option in `args-token` validation
 * @returns A rendered validation error.
 */
export function renderValidationErrors<G extends GunshiParams = DefaultGunshiParams>(
  _ctx: CommandContext<G>,
  error: AggregateError
): Promise<string> {
  const messages = [] as string[]
  for (const err of error.errors as Error[]) {
    messages.push(err.message)
  }
  return Promise.resolve(messages.join('\n'))
}
