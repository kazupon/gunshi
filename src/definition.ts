/**
 * The entry for command deifinition.
 *
 * @example
 * ```js
 * import { define } from 'gunshi/definition'
 * ```
 *
 * @module
 */

/**
 * @author kazuya kawaguchi (a.k.a. kazupon)
 * @license MIT
 */

import type { ArgOptions } from 'args-tokens'
import type { Command, CommandLoader, LazyCommand } from './types.ts'

export type { ArgOptions, ArgOptionSchema, ArgValues } from 'args-tokens'

/**
 * Define a {@link Command | command} with type inference
 * @param definition A {@link Command | command} definition
 * @returns A {@link Command | command} definition with type inference
 */
export function define<Options extends ArgOptions = ArgOptions>(
  definition: Command<Options>
): Command<Options> {
  return definition
}

/**
 * Define a {@link LazyCommand | lazy command} with command loader, which is attached with command definition as usage metadata.
 * @param loader A {@link CommandLoader | command loader}
 * @param definition A {@link Command | command} definition
 * @returns A {@link LazyCommand | lazy command} loader
 */
export function lazy<Options extends ArgOptions = ArgOptions>(
  loader: CommandLoader<Options>,
  definition?: Command<Options>
): LazyCommand<Options> {
  if (definition != null) {
    ;(loader as LazyCommand<Options>).commandName = definition.name
    ;(loader as LazyCommand<Options>).description = definition.description
    ;(loader as LazyCommand<Options>).options = definition.options
    ;(loader as LazyCommand<Options>).examples = definition.examples
    ;(loader as LazyCommand<Options>).resource = definition.resource
  }
  return loader
}
