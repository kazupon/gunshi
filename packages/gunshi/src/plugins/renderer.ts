/**
 * @author kazuya kawaguchi (a.k.a. kazupon)
 * @license MIT
 */

import { plugin } from '../plugin.ts'
import { renderHeader } from '../renderer/header.ts'
import { renderUsage } from '../renderer/usage.ts'
import { renderValidationErrors } from '../renderer/validation.ts'

/**
 * Default renderer plugin for Gunshi.
 */
export default function renderer() {
  return plugin({
    name: 'default-renderer',
    setup: ctx => {
      ctx.decorateHeaderRenderer(async (_baseRenderer, cmdCtx) => await renderHeader(cmdCtx))
      ctx.decorateUsageRenderer(async (_baseRenderer, cmdCtx) => await renderUsage(cmdCtx))
      ctx.decorateValidationErrorsRenderer(
        async (_baseRenderer, cmdCtx, error) => await renderValidationErrors(cmdCtx, error)
      )
    }
  })
}
