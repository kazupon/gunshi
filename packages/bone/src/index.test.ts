import { define } from '@gunshi/definition'
import global from '@gunshi/plugin-global'
import renderer from '@gunshi/plugin-renderer'
import { describe, expect, test, vi } from 'vitest'
import { Args, cli, GunshiParams, RemovedIndex } from './index.ts'

type GlobalsCommandContext = Awaited<ReturnType<ReturnType<typeof global>['extension']['factory']>>

// TODO(kazupon): more useful type inference for exntensions
const entry = define<
  GunshiParams<{ args: Args; extensions: { 'g:globals': RemovedIndex<GlobalsCommandContext> } }>
>({
  name: 'entry',
  args: {
    say: {
      type: 'string',
      description: 'say something',
      default: 'hello!'
    }
  },
  run: ctx => {
    if (ctx.values.say === 'version?') {
      ctx.extensions['g:globals'].showVersion()
      return
    }
    return `You said: ${ctx.values.say}`
  }
})

const meta = {
  name: 'app-bone',
  description: 'app-bone is a @gunshi/bone example',
  version: '1.0.0'
}

describe('@gunshi/bone', async () => {
  test('conbination with @gunshi/definition', async () => {
    const result = await cli(['--say', '`hello world`'], entry, { ...meta })
    expect(result).toEqual('You said: `hello world`')
  })

  test('conbination with gunshi plugins', async () => {
    const logs: string[] = []
    vi.spyOn(console, 'log').mockImplementation((msg: string) => logs.push(msg))

    await cli(['-h'], entry, {
      ...meta,
      plugins: [global(), renderer()]
    })
    expect(logs.join('\n')).toMatchSnapshot()
  })

  test('conbination with gunshi plugins and extension', async () => {
    const logs: string[] = []
    vi.spyOn(console, 'log').mockImplementation((msg: string) => logs.push(msg))

    await cli(['--say', 'version?'], entry, {
      ...meta,
      plugins: [global(), renderer()]
    })
    expect(logs.join('\n')).toMatchSnapshot()
  })
})
