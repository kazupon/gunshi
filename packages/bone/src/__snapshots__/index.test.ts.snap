// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@gunshi/bone > conbination with gunshi plugins 1`] = `
"app-bone is a @gunshi/bone example (app-bone v1.0.0)

USAGE:
  app-bone <OPTIONS>

OPTIONS:
  -h, --help             Display this help message
  -v, --version          Display this version
  --say [say]            say something (default: hello!)
"
`;

exports[`@gunshi/bone > conbination with gunshi plugins and extension 1`] = `
"app-bone is a @gunshi/bone example (app-bone v1.0.0)

1.0.0"
`;
